## About `ideal`

`ideal` is a Bioconductor package containing a Shiny application for
interactively analyzing RNA-seq expression data, by interactive exploration of the 
results of a Differential Expression analysis.

`ideal` 

Thanks to its interactive/reactive design, it is designed to become a practical
companion to any RNA-seq dataset analysis, making downstream and exploratory 
data analysis accessible also to the bench biologist, while providing additional 
quick insight also for the experienced data analyst.

`ideal` was developed in the Bioinformatics Division led by Harald Binder 
at the IMBEI (Institut f√ºr Medizinische Biometrie, Epidemiologie und Informatik) 
in the University Medical Center of the Johannes Gutenberg University Mainz.



## Developers

<a href="mailto:marinif@uni-mainz.de?subject=[pcaExplorer question]" class="btn btn-primary">Federico Marini</a>

## Code

All code for `ideal` is available on 
<a href="https://github.com/federicomarini/ideal" target="_blank">GitHub</a>.


## Citation info

If you use `ideal` for your analysis, please cite it as here below:

...
